### ISSUE TITLE
format on save no longer works in 0.0.20

### ISSUE THREAD

**Issue by cludden** [mauve/vscode-terraform#45]
I can see tflint running in the output, but it is no longer formatting the code on save

**Comment by brandoconnor** [comment]
Also getting this behavior under 0.0.20. VScode details:

![screen shot 2017-12-18 at 16 52 34](https://user-images.githubusercontent.com/2502520/34135192-e302d176-e413-11e7-9ffd-f6a4106594a4.png)

**Comment by brandoconnor** [comment]
As a workaround, I figured out how to install the previous version:

[This stackoverflow answer](https://stackoverflow.com/questions/37071388/how-to-install-vscode-extensions-offline) guides you through the process. The link to download the [previous version is here](https://mauve.gallery.vsassets.io/_apis/public/gallery/publisher/mauve/extension/terraform/0.0.19/assetbyname/Microsoft.VisualStudio.Services.VSIXPackage).

**Comment by jacobwgillespie** [comment]
It looks like the `terraform.formatOnSave` setting was removed in favor of using the built-in `editor.formatOnSave` setting.  If you want to enable format-on-save only for Terraform files, you can add the following to your user settings or workspace settings:

```json
{
  "[terraform]": {
    "editor.formatOnSave": true
  }
}
```

**Comment by jacobwgillespie** [comment]
Actually the changelog mentions this: https://github.com/mauve/vscode-terraform/blob/master/CHANGELOG.md#0020.

**Comment by brandoconnor** [comment]
I should have been more clear, @jacobwgillespie . That config was tried, the old style was removed, and the Terraform output console shows `terraform fmt` successfully running against files that continue to not get formatted. I think there's still a bug here.

```
terraform.format: running 'terraform fmt' on '/Users/brandonoconnor/Development/git/foo/variables.tf'
terraform.format: Successful.
```

**Comment by jacobwgillespie** [comment]
Ah, I see, gotcha.  Well in that case, I can confirm that format-on-save is working for me with the above settings, vscode-terraform version 0.0.20, and VS Code 1.19.0.  Do you have any other settings or extensions that might influence file formatting?

**Comment by mauve** [comment]
I just tried to reproduce this but cannot reproduce for `editor.formatOnSave`. Format on save works for me with the following setting:

```
"editor.formatOnSave": true
```

or

```
{
  "[terraform]": {
    "editor.formatOnSave": true
  }
}
```

However format on save does *not* work with `editor.autoSave` which I think is *by design* in VSCode, at least this is what the settings documentation says:

```
// Format a file on save. A formatter must be available, the file must not be auto-saved, and editor must not be shutting down.
"editor.formatOnSave": false,
```

The formatter does not even seem to be invoked when `editor.autoSave` is `true`, however manually invoking save by `Cmd+S` or `Ctrl+S`.

**Comment by mauve** [comment]
@cludden: just a quick _fyi_: `tflint` is for linting, it has nothing with formatting to do. Formatting is done by `terraform fmt`, you should see the following output in the Output tab:

```
terraform.format: running 'terraform fmt' on '/Users/olenfalk/projects/vscode-terraform/templates/template.tf'
terraform.format: Successful.
```

**Comment by mauve** [comment]
@brandoconnor @cludden just to be clear do you have problems with the `formatOnSave` in general or is it the removal/breakage of the `formatOnAutoSave` feature which you are referring to?

**Comment by mauve** [comment]
Format on save is disabled when the save reason is auto:

https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/api/electron-browser/mainThreadSaveParticipant.ts#L190-L193

**Comment by cludden** [comment]
@mauve I wasn't aware of the `editor.formatOnSave` setting.. that fixed it. thanks for clearing it up!

**Comment by brandoconnor** [comment]
@cludden - Nope still not fixed here. Please reopen.

@mauve - I've verified that `"editor.formatOnSave": true`, `editor.autoSave` is non-existent (not even a valid configuration), and formatting is working on other file types. As mentioned in my previous comment, `terraform.format` shows the expected output on the correct file but nothing happens to the file. If I go to the terminal and `terraform fmt` manually, the file is formatted, but the extension is still not doing what it claims to be doing in the output. Any other settings to try or debugging that can be turned up?

**Comment by jgrossmac** [comment]
broken for me as well...

**Comment by mauve** [comment]
@brandoconnor ok I will take another look tonight, sorry for breaking your setup, if you can think of any other info which might narrow this down for me please keep commenting here.

**Comment by brandoconnor** [comment]
Just tried once more with both settings as suggested on #46 and that wasn't successful.

**Comment by taganaka** [comment]
Seems like uninstalling and re-installing the extension helps

**Comment by brandoconnor** [comment]
Wish I could have that same story @mspaulding06 🤷‍♀️ . I've probably uninstalled and reinstalled the new version 10x in the course of troubleshooting.

**Comment by tommynsong** [comment]
Still doesn't work for me, i have re-installed the plugins several time and also set both `"editor.formatOnSave": true` and `"terraform.format.enable": true`

**Comment by mauve** [comment]
Hi guys, thanks for being patient:

Please test the new version and tell me if it still doesn't work for you:


0.0.21
======

## What's New
- Auto completion support [@ranga543](https://github.com/ranga543)
- Format on save is back (closes #45, #47)
- Tools can now be installed in directories which have spaces [@xeres](https://github.com/xeres)

### Format on save

The last version introduced usage of the correct extension APIs for registering formatting
providers. VSCode then automatically invokes the formatting provider on save if `editor.formatOnSave` has been specified. VSCode however does never invoke the formatting
provider if the save operation came from an auto-save event.

0.0.21 reintroduces the manual code which handled format on autosave instead of relying on
VSCode to perform those operations. The following configuration options have changes as result
of that:

- `terraform.format.enable`: needs to be `true`
- `terraform.format.formatOnSave`: needs to be `true`
- `terraform.format.ignoreExtensionsOnSave`: can be used to exclude `.tfvars` for example from autosave formatting

The global setting: `editor.formatOnSave` should be `false` for Terraform files and the
extension contributes that default configuration. Please verify that you do not have
conflicting configuration, by removing the following setting if you have it:

```
"[terraform]": {
  "editor.formatOnSave": true
}
```

The configuration `editor.formatOnSave` will have the correct setting contributed by Terraform.
